reactive/gcp.py

```diff
@@ -10,16 +10,22 @@
     hook,
 )
 from charms.reactive.relations import endpoint_from_name
+import charms.layer.snap
 from charmhelpers.core import hookenv
 
 from charms import layer
 
 
 @when_all("snap.installed.google-cloud-sdk")
+def remove_snap():
+    charms.layer.snap.remove("google-cloud-sdk")
```

Comment:
presumably this removes the `snap.installed.google-cloud-sdk` flag?

Reply:
indeed [it does](https://github.com/charmed-kubernetes/layer-snap/blob/main/lib/charms/layer/snap.py#L123-L126)

Reply:
This needs to be removed *before* the google-cloud-cli snap can be installed, right? Because of this?

```
$ sudo snap install google-cloud-sdk --classic
google-cloud-sdk 400.0.0 from Cloud SDK (google-cloud-sdk✓) installed

$ sudo snap install google-cloud-cli --classic
error: cannot perform the following tasks:
- Set automatic aliases for snap "google-cloud-cli" (cannot enable aliases "gcloud", "gsutil", "anthoscli", "docker-credential-gcloud", "bq" for "google-cloud-cli", already enabled for "google-cloud-sdk")
```

If so, then we have problem with the order of operations on upgrade. The snap layer registers a hookenv.atstart install function [here](https://github.com/charmed-kubernetes/layer-snap/blob/a0e1f28e8bb9040eada9a7a73f66ee6a615704b7/reactive/snap.py#L341) that will try to install the new google-cloud-cli snap [here](https://github.com/charmed-kubernetes/layer-snap/blob/a0e1f28e8bb9040eada9a7a73f66ee6a615704b7/reactive/snap.py#L82-L83). That's guaranteed to run before this `remove_snap` handler has had a chance to run, so I think on an upgraded deployment, that install is going to fail.

Reply:
I don't think we could easily inject a call to snap.remove before layer-snap tries to install all its snaps. To get around that, I would remove the google-cloud-* snap from layer.yaml to prevent layer-snap from trying to install it at all. Then we can install it ourselves by calling `charms.layer.snap.install(...)`.

Reply:
how about a change to layer-snap to account for this?

I could add a new option for it to support to list any snaps that need to be removed:
```yaml
  snap:
    google-cloud-sdk:
       remove: null
    google-cloud-cli:
      channel: stable
      classic: true
```

Reply:
* https://github.com/charmed-kubernetes/layer-snap/pull/2